flow/project.py

```diff
@@ -267,6 +268,25 @@ def _make_bundles(operations, size=None):
             break
 
 
+class _HooksDecorator:
+    def __init__(self):
+        self._operation_hooks = defaultdict(lambda: defaultdict(list))
+
+    def __getitem__(self, func):
+        return self._operation_hooks[func]
+
```

Comment:
Should we define this allowing users to access the current hooks through `func_hooks = FlowProject.hook[func]`? We don't expose `FlowProject._operation_hooks` but do expose `_HookDecorator`'s hook list. This means that after instantiation `hooks = project.hook[op_func]` may not be the true hooks used, and does not point to the true store of that information at that point `project._operation_hooks`.

Reply:
@klywang and I agreed this should be removed since it allows wacky and technically user facing (no underscored methods needed) behavior.

Reply:
@b-butler I just checked, and it turns out `hooks = project.hook[op_func]` returns an empty list rather than `_HookRegistry`'s hook list. In order to access the actual hooks, you need to access an underscored method (i.e.  `project.hook._operation_hooks[foo]['on_start']`). I guess that's a little safer since you can't do `project.hook[op_func].on_start.append(foo)` without error. However, it's still possible to do something like `project.hook[op_func]['on_start'].append(foo)`. 

Also, other attribute for `_FlowProjectClass`, like `pre`, `post`, or `operation`, are set using static methods  (lines 1248 - 1388, static methods with naming like `_setup_*`). For the sake of consistency, should we move `_HooksRegistry` to a new static method in `_FlowProjectClass`?

Reply:
This shouldn't be the case since `_HookRegistry.__getitem__` just returns the internal `_operation_func` value. Regardless, I don't think this magic method is necessary, and is better of removed.